src/traits.rs

```diff
@@ -116,15 +116,64 @@ impl Lerp for f32 {
     }
 }
 
-use crate::color::{LSample, LaSample, RgbSample, RgbaSample};
+pub trait Sample: Primitive + 'static {
+    const MAX_SAMPLE_VALUE: Self;
+    const RGB_COLOR_TYPE: ColorType;
+    const RGBA_COLOR_TYPE: ColorType;
+    const L_COLOR_TYPE: ColorType;
+    const LA_COLOR_TYPE: ColorType;
```

Comment:
Perhaps these should be `Option<ColorType>`?

Reply:
I tried that, but I wasn't brave enough to also make the `trait Pixel::COLOR_TYPE` optional. That would be a lot cleaner though. And it would result in runtime errors instead of subtly incorrect behavior.

Reply:
I can do that. Alternatively, we could also make a separate trait only for the color type. Then we could rule out these errors at compile time, instead of sometimes throwing a runtime error.

```rust
trait PixelWithColorType {
    const COLOR_TYPE: ColorType;
}
```

And then only some pixels will implement that, and the methods that use this constants could require this trait in addition to `Pixel`. What do you think?

Reply:
I'd like to do this:
```rust
impl Sample for f32 {
    const RGBA_COLOR_TYPE: ImageResult<ColorType> = Ok(ColorType::Rgba32F);
    const L_COLOR_TYPE: ImageResult<ColorType> = Err(UnsupportedPixelError);
}

/*const UnsupportedPixelError: ImageError = ImageError::Unsupported(UnsupportedError::from_format_and_kind(
    ImageFormatHint::Unknown, UnsupportedErrorKind::Color(ExtendedColorType::Unknown(255))
));*/

/*const UnsupportedPixelError: ImageError = ImageError::Parameter(ParameterError::from_kind(
    ParameterErrorKind::FailedAlready
));*/
```

But I can't find an error variant which would make sense

Reply:
I'm nervous that fracturing Pixel like that. Downstream code might end up having to carry around a whole bunch of different trait bounds. Worse adding any new bounds would be API breaking.

Making Pixel::COLOR_TYPE optional would be an improvement. Though I wonder whether it would be feasible to simply only implement Pixel for color types. That might entail converting `Rgb<T>` -> `Rgb8` + `Rgb16` + `Rgb32F`, etc.

Reply:
> Worse adding any new bounds would be API breaking

How is that worse than changing the associated constant in a trait? 

In general, I would not say that a monolithic trait has advantages. The pixel trait has multiple responsibilities, two of which are (a) data storage for image processing and (b) color format inference for image file saving. Having both of those responsibilities in one trait is problematic, because you will end up in one of the following drawbacks: Either you force a color format on all pixels (even if you only need and intermediate `Luma<f32>` image), which now is a runtime error. Or you unnecessarily limit color processing algorithms and temporary storage to the few select pixel types which can be saved to a file, which could otherwise also work with `Luma<f32>`.

Reply:
I got a little carried away, thinking of an ideal solution. To cut this short, I think that splitting up the trait may be a larger change, for another time. Maybe we should focus on going `f32` with the smallest amount of changes for now :)